src/main_gate.rs

```diff
@@ -360,31 +366,31 @@ impl<F: PrimeField, const T: usize> MainGate<F, T> {
             v2.clone() * v2 * v
         };
 
-        meta.create_gate("q_m*s[0]*s[1] + sum_i(q_1[i]*s[i]) + sum_i(q_5[i]*s[i]^5) + rc + q_i*input + q_o*out=0", |meta|{
+        meta.create_gate("q_m[0]*s[0]*s[1] + q_m[1]*s[2]*s[3] + sum_i(q_1[i]*s[i]) + sum_i(q_5[i]*s[i]^5) + rc + q_i*input + q_o*out=0", |meta|{
             let state = state.into_iter().map(|s| meta.query_advice(s, Rotation::cur())).collect::<Vec<_>>();
             let input = meta.query_advice(input, Rotation::cur());
             let out = meta.query_advice(out, Rotation::cur());
             let q_1 = q_1.into_iter().map(|q| meta.query_fixed(q, Rotation::cur())).collect::<Vec<_>>();
             let q_5 = q_5.into_iter().map(|q| meta.query_fixed(q, Rotation::cur())).collect::<Vec<_>>();
-            let q_m = meta.query_fixed(q_m, Rotation::cur());
+            let q_m = q_m.into_iter().map(|q| meta.query_fixed(q, Rotation::cur())).collect::<Vec<_>>();
             let q_i = meta.query_fixed(q_i, Rotation::cur());
             let q_o = meta.query_fixed(q_o, Rotation::cur());
             let rc = meta.query_fixed(rc, Rotation::cur());
 
-            let init_term = q_m * state[0].clone() * state[1].clone() + q_i * input + rc + q_o * out;
+            let init_term = match T {
+                0..=1 => unreachable!("assert with `T >= 2` before"),
+                2..=3 => q_m[0].clone() * state[0].clone() * state[1].clone() + q_i * input + rc + q_o * out,
+                _ => q_m[1].clone() * state[2].clone() * state[3].clone() + q_m[0].clone() * state[0].clone() * state[1].clone() + q_i * input + rc + q_o * out,
+            };
```

Comment:
The order in which expression is calculated can be meaningful, so I kept it. Let me know if I've overdone it, then this code could be made more concise

Reply:
You mean the order with 0..=1 first, 2..=2 second?

Reply:
No, I can do same code shorter
```rust
let mut init_term = q_m[0].clone() * state[0].clone() * state[1].clone() + q_i * input + rc + q_o * out;
                                                                                                         
if T >= 4 {
    init_term = q_m[1].clone() * state[2].clone() * state[3].clone() + init_term;
}
```
but order of operations with expressions will be different

Reply:
I've been digging through the source code, I think I've overdone it and there's a simpler way to do it here.

Reply:
The previous one can handle the `T<2` case correctly.  With this new modification, it will always return init_term even in the case `T<2`, which we want to return error. Unless, we enforce to return T>=2

Reply:
We previously have an assert that does not allow T = 0 or 1. The previous version panicked about this as well (`unreachable!`). Current one will panic at `self.state[1]`.

We can add support for such sizes, but I suggest to do it in a separate PR to also include a separate test about it.